File: reader/DNASignBaseline.cpp
// warning: this is extremely broken atm
// a lot of debug stuff have been left behind

#include "DNASignBaseline.h"
#include <nodePath.h>
#include <textNode.h>
#include <nodePathCollection.h>
#include <compress_string.h>
#include <bamFile.h>

#define M_PI 3.14159265358979323846  /* pi */

#include "DNASignText.h"

DNASignBaseline::DNASignBaseline(string name) : DNANode(name), m_data("") {}
DNASignBaseline::~DNASignBaseline(void) {}

void DNASignBaseline::make_from_dgi(DatagramIterator &dgi, DNAStorage *store)
{
	DNANode::make_from_dgi(dgi, store);

	string comp_string = dgi.get_string();
	if (comp_string.size() > 0)
	{
		m_data = decompress_string(comp_string);
	};
};

void DNASignBaseline::traverse(NodePath& np, DNAStorage* store)
{
	NodePath _np = np.attach_new_node("baseline");
	_np.set_pos(*m_pos);
	_np.set_hpr(*m_hpr);
	_np.set_depth_offset(50);

	stringstream ss;
	ss << m_data;

	if (m_data.size() > 0)
	{
		BamFile bf = BamFile();
		bf.open_read(ss);
		NodePath node = NodePath(bf.read_node());
		node.reparent_to(_np);
	};

	for (comp_vec_t::iterator child = m_children.begin(); child != m_children.end(); child++)
	{
		(*child)->traverse(_np, store);
	};
};

File: libpandadna/DNASignBaseline.py
import DNANode

from common import *

class DNASignBaseline(DNANode.DNANode):
    COMPONENT_CODE = 6

    def __init__(self):
        DNANode.DNANode.__init__(self, '')

        self.data = ""
        
    def makeFromDGI(self, dgi):
        DNANode.DNANode.makeFromDGI(self, dgi)
        
        self.data = dgi.getString()
        if len(self.data):
            self.data = decompressString(self.data)
            
    def traverse(self, np, store):
        _np = np.attachNewNode("baseline")
        _np.setPos(self.pos)
        _np.setHpr(self.hpr)
        _np.setDepthOffset(50)

        if self.data:
            bf = BamFile()
            ss = StringStream()
            ss.setData(self.data)
            bf.openRead(ss)
            
            node = NodePath(bf.readNode())
            node.reparentTo(_np)

        self.traverseChildren(_np, store)
        